### PR TITLE
set_last_validated_era + tests

### PR DESCRIPTION
**Pull request by aie0**
## Types of Changes
Please select the branch type you are merging and fill in the relevant template.

- [ ] Hotfix
- [ ] Release
- [ ] Fix or Feature

### Types of Changes

- [ ] Tech Debt (Code improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Dependency upgrade (A change in substrate or any 3rd party crate version)

### Migrations and Hooks

- [x] This change requires a runtime migration.
- [ ] Modifies `on_initialize`
- [ ] Modifies `on_finalize`

### Checklist for Fix or Feature

- [x] Change has been tested locally.
- [x] Change adds / updates tests if applicable.
- [ ] Changelog doc updated.
- [ ] `spec_version` has been incremented.
- [ ] `network-relayer`'s [events](https://github.com/Cerebellum-Network/network-relayer/blob/dev-cere/shared/substrate/events.go) have been updated according to the blockchain events if applicable.
- [ ] All CI checks have been passed successfully

## Checklist for Hotfix

- [ ] Changelog has been updated.
- [ ] Crate version has been updated.
- [ ] `spec_version` has been incremented.
- [ ] Transaction version has been updated if required.
- [ ] Pull Request to `dev` has been created.
- [ ] Pull Request to `staging` has been created.
- [ ] `network-relayer`'s [events](https://github.com/Cerebellum-Network/network-relayer/blob/dev-cere/shared/substrate/events.go) have been updated according to the blockchain events if applicable.
- [ ] All CI checks have been passed successfully

## Checklist for Release

- [ ] Change has been deployed to Devnet.
- [ ] Change has been tested in Devnet.
- [ ] Change has been deployed to Qanet.
- [ ] Change has been tested in Qanet.
- [ ] Change has been deployed to Testnet.
- [ ] Change has been tested in Testnet.
- [ ] Changelog has been updated.
- [ ] Crate version has been updated.
- [ ] Spec version has been updated.
- [ ] Transaction version has been updated if required.
- [ ] All CI checks have been passed successfully

### CODE REVIEW

File: pallets/ddc-clusters/src/lib.rs [hunk 1]
```diff
@@ -815,6 +816,23 @@ pub mod pallet {
 	}
 
 	impl<T: Config> ClusterManager<T> for Pallet<T> {
+		fn set_last_validated_era(
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/lib.rs:819]: I think we need to create a dedicated trait, i.e. `ClusterValidator`, and put the `set_last_validated_era` and other validation-related methods there.

The `ClusterManager` trait is supposed to contain functions related to DDC nodes management
The `ClusterProtocol` trait is supposed to contain functions related to Cluster economics
The `ClusterValidator` trait will contain validation-related functions. 

By keeping those traits separately, you decrease the chance that someone will call your `set_last_validated_era` except the `ddc-payouts` pallet.

It also makes the design clearer and the code more readable.

**Review by yahortsaryk — CHANGES_REQUESTED** [review]

File: pallets/ddc-clusters/src/migrations.rs [hunk 2]
```diff
@@ -402,3 +405,151 @@ pub mod v2 {
...
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+			frame_support::ensure!(
+				Pallet::<T>::on_chain_storage_version() == 1,
+				"must upgrade linearly"
+			);
+			let pre_clusters_count = Clusters::<T>::iter_keys().count();
+			let pre_clusters_nodes_count = ClustersNodes::<T>::iter_keys().count();
+			let pre_clusters_nodes_stats_count = ClustersNodesStats::<T>::iter_keys().count();
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/migrations.rs:500]: Why do you need to check for `pre_clusters_nodes_stats_count` and `pre_clusters_nodes_count` if your migration doesn't do anything to the `ClustersNodesStats` and `ClustersNodes` storage items ?

File: pallets/ddc-clusters/src/migrations.rs [hunk 3]
```diff
@@ -402,3 +405,151 @@ pub mod v2 {
...
+	pub type OldNodeStatus = bool;
+	pub struct MigrateToV3<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV3<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let current_version = Pallet::<T>::current_storage_version();
+			let onchain_version = Pallet::<T>::on_chain_storage_version();
+			let mut weight = T::DbWeight::get().reads(1);
+
+			if onchain_version == 2 && current_version == 3 {
+				let mut nodes_count_by_cluster: BTreeMap<ClusterId, ClusterNodesCount> =
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/migrations.rs:454]: I don't see usages of the `nodes_count_by_cluster` variable, I think we should remove it

File: pallets/ddc-clusters/src/migrations.rs [hunk 4]
```diff
@@ -402,3 +405,151 @@ pub mod v2 {
...
+			frame_support::ensure!(
+				Pallet::<T>::on_chain_storage_version() == 1,
+				"must upgrade linearly"
+			);
+			let pre_clusters_count = Clusters::<T>::iter_keys().count();
+			let pre_clusters_nodes_count = ClustersNodes::<T>::iter_keys().count();
+			let pre_clusters_nodes_stats_count = ClustersNodesStats::<T>::iter_keys().count();
+
+			assert_eq!(
+				pre_clusters_nodes_stats_count, 0,
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/migrations.rs:503]: This check will fail on Devnet/Qanet/Tetstnet/Mainnet as clusters on all environments have the related `ClustersNodeStats` item

File: pallets/ddc-clusters/src/migrations.rs [hunk 5]
```diff
@@ -402,3 +405,151 @@ pub mod v2 {
...
+			);
+			let post_clusters_nodes_count = ClustersNodes::<T>::iter().count() as u32;
+			assert_eq!(
+				pre_clusters_nodes_count, post_clusters_nodes_count,
+				"the clusters nodes count before (pre: {}) and after (post: {})  the migration should be the same", 
+				pre_clusters_nodes_count, post_clusters_nodes_count
+			);
+
+			let post_clusters_nodes_stats_count = ClustersNodesStats::<T>::iter().count() as u32;
+			assert_eq!(
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/migrations.rs:531]: Same here, you shouldn't check for `post_clusters_nodes_count` and `post_clusters_nodes_count` if you don't modify the related storage items

**Review by yahortsaryk — DISMISSED** [review]

File: runtime/cere-dev/src/lib.rs [hunk 6]
```diff
@@ -1254,6 +1254,7 @@ impl pallet_ddc_verification::Config for Runtime {
 	type PalletId = VerificationPalletId;
 	type WeightInfo = pallet_ddc_verification::weights::SubstrateWeight<Runtime>;
 	type ClusterManager = pallet_ddc_clusters::Pallet<Runtime>;
+	type ClusterValidator = pallet_ddc_clusters::Pallet<Runtime>;
 	type NodeVisitor = pallet_ddc_nodes::Pallet<Runtime>;
 	type PayoutVisitor = pallet_ddc_payouts::Pallet<Runtime>;
 	type AuthorityId = ddc_primitives::sr25519::AuthorityId;
```
Comment (yahortsaryk) [inline runtime/cere-dev/src/lib.rs:1260]: You should include your `pallet_ddc_clusters::migration::MigrateToV3<Runtime>` to the `Migrations` tuple of the runtime,  i.e. there should be the entry of ```type Migrations = (pallet_ddc_clusters::migration::MigrateToV3<Runtime>)``` in the `runtime/cere-dev/src/lib.rs` file

File: node/service/chain-specs/example.json [hunk 7]
```diff
@@ -95,6 +95,36 @@
           ]
         ]
       },
+      "democracy": {
```
Comment (yahortsaryk) [inline node/service/chain-specs/example.json:98]: The `democracy` pallet was removed when we introduced `OpenGov`, you should remove it from the `example.json` file.

File: node/service/chain-specs/example.json [hunk 8]
```diff
@@ -95,6 +95,36 @@
...
+        "phantom": null
+      },
+      "council": {
```
Comment (yahortsaryk) [inline node/service/chain-specs/example.json:101]: The `council` pallet was removed when we introduced `OpenGov`, you should remove it from the `example.json` file.

File: node/service/chain-specs/example.json [hunk 9]
```diff
@@ -95,6 +95,36 @@
...
+        "phantom": null,
+        "members": []
+      },
+      "technicalCommittee": {
```
Comment (yahortsaryk) [inline node/service/chain-specs/example.json:105]: The `technicalCommittee` pallet was removed when we introduced `OpenGov`, you should remove it from the `example.json` file. We have `techComm` now.

File: node/service/chain-specs/example.json [hunk 10]
```diff
@@ -95,6 +95,36 @@
...
+        "phantom": null,
+        "members": [
+          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
+          "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"
+        ]
+      },
+      "elections": {
```
Comment (yahortsaryk) [inline node/service/chain-specs/example.json:112]: The `elections` pallet was removed when we introduced OpenGov, you should remove it from the example.json file.

File: node/service/chain-specs/example.json [hunk 11]
```diff
@@ -95,6 +95,36 @@
...
+            "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
+            10000000000
+          ],
+          [
+            "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
+            10000000000
+          ]
+        ]
+      },
+      "technicalMembership": {
```
Comment (yahortsaryk) [inline node/service/chain-specs/example.json:124]: The `technicalMembership` pallet was removed when we introduced OpenGov, you should remove it from the example.json file.

File: node/service/chain-specs/example.json [hunk 12]
```diff
@@ -108,6 +138,14 @@
       "authorityDiscovery": {
         "keys": []
       },
+      "society": {
```
Comment (yahortsaryk) [inline node/service/chain-specs/example.json:141]: The `society"` pallet was removed previously. You should remove it from the example.json file.

File: pallets/ddc-clusters/src/migrations.rs [hunk 13]
```diff
@@ -402,3 +419,185 @@ pub mod v2 {
...
+			let post_count = Clusters::<T>::iter().count() as u64;
+			ensure!(
+				prev_count == post_count,
+				"the cluster count before and after the migration should be the same"
+			);
+
+			let current_version = Pallet::<T>::current_storage_version();
+			let on_chain_version = Pallet::<T>::on_chain_storage_version();
+
+			ensure!(current_version == 3, "must_upgrade");
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/migrations.rs:511]: In `post_upgrade` you can additionally check what your migrations modified, i.e. you can check that `last_validated_era_id` field is set to 0 for all clusters and there is no clusters with other values.

File: runtime/cere-dev/src/lib.rs [hunk 14]
```diff
@@ -1401,7 +1403,13 @@ pub type SignedPayload = generic::SignedPayload<RuntimeCall, SignedExtra>;
 pub type CheckedExtrinsic = generic::CheckedExtrinsic<AccountId, RuntimeCall, SignedExtra>;
 
 /// Runtime migrations
-type Migrations = ();
+type Migrations = (
+	pallet_ddc_clusters::migrations::v3::MigrateToV3<Runtime>,
+	pallet_ddc_clusters::migrations::v2::MigrateToV2<Runtime>,
```
Comment (yahortsaryk) [inline runtime/cere-dev/src/lib.rs:1408]: `pallet_ddc_clusters::migrations::v2` has been executed already in `cere_dev`. You should remove it from the list.

File: runtime/cere-dev/src/lib.rs [hunk 15]
```diff
@@ -1401,7 +1403,13 @@ pub type SignedPayload = generic::SignedPayload<RuntimeCall, SignedExtra>;
...
+	pallet_ddc_staking::migrations::v1::MigrateToV1<Runtime>,
```
Comment (yahortsaryk) [inline runtime/cere-dev/src/lib.rs:1409]: `pallet_ddc_staking::migrations::v1` has been executed already in `cere_dev`. You should remove it from the list.

File: runtime/cere-dev/src/lib.rs [hunk 16]
```diff
@@ -1401,7 +1403,13 @@ pub type SignedPayload = generic::SignedPayload<RuntimeCall, SignedExtra>;
...
+	pallet_ddc_customers::migration::MigrateToV2<Runtime>,
```
Comment (yahortsaryk) [inline runtime/cere-dev/src/lib.rs:1410]: `pallet_ddc_customers::migration::MigrateToV2` has been executed already in `cere_dev`. You should remove it from the list.

File: runtime/cere-dev/src/lib.rs [hunk 17]
```diff
@@ -1401,7 +1403,13 @@ pub type SignedPayload = generic::SignedPayload<RuntimeCall, SignedExtra>;
...
+	pallet_ddc_customers::migration::MigrateToV1<Runtime>,
```
Comment (yahortsaryk) [inline runtime/cere-dev/src/lib.rs:1411]: `pallet_ddc_customers::migration::MigrateToV1` has been executed already in `cere_dev`. You should remove it from the list.

File: runtime/cere/constants/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,78 @@
+// This file is part of Substrate.
```
Comment (yahortsaryk) [inline runtime/cere/constants/src/lib.rs:1]: This file (`runtime/cere/constants/src/lib.rs`) should be completely removed. We keep constants in `runtime/common/src/constants.rs` file now. 
The `runtime/cere/constants/src/lib.rs` is out of use and is not imported by any crate currently.

File: runtime/cere-dev/src/lib.rs [hunk 19]
```diff
@@ -501,6 +501,7 @@ impl_opaque_keys! {
 		pub babe: Babe,
```
Comment (yahortsaryk) [inline runtime/cere-dev/src/lib.rs:501]: To nee to increment the `spec_version` in `cere_dev` runtime.

File: runtime/cere/src/lib.rs [hunk 20]
```diff
@@ -1303,6 +1303,7 @@ impl pallet_ddc_verification::Config for Runtime {
 	type PalletId = VerificationPalletId;
```
Comment (yahortsaryk) [inline runtime/cere/src/lib.rs:1303]: To nee to increment the `spec_version` in `cere` runtime.

File: runtime/cere/src/lib.rs [hunk 21]
```diff
@@ -1410,9 +1411,11 @@ pub type CheckedExtrinsic = generic::CheckedExtrinsic<AccountId, RuntimeCall, Si
 
 /// Runtime migrations
 type Migrations = (
+	pallet_ddc_clusters::migrations::v3::MigrateToV3<Runtime>,
 	pallet_ddc_clusters::migrations::v2::MigrateToV2<Runtime>,
 	pallet_ddc_staking::migrations::v1::MigrateToV1<Runtime>,
 	pallet_ddc_customers::migration::MigrateToV2<Runtime>,
+	pallet_ddc_customers::migration::MigrateToV1<Runtime>,
```
Comment (yahortsaryk) [inline runtime/cere/src/lib.rs:1418]: `pallet_ddc_customers::migration::MigrateToV1` has been executed on Mainnet. You should remove it from the list.

File: runtime/cere/src/lib.rs [hunk 22]
```diff
@@ -1410,9 +1411,11 @@ pub type CheckedExtrinsic = generic::CheckedExtrinsic<AccountId, RuntimeCall, Si
...
 	migrations::Unreleased,
```
Comment (yahortsaryk) [inline runtime/cere/src/lib.rs:1419]: You should see if any of `migrations::Unreleased` migrations have been applied to Mainnet. If yes, you should remove it from the list.

**Review by yahortsaryk — CHANGES_REQUESTED** [review]

**Comment by aie0** [comment]
obsolete in favour of https://github.com/Cerebellum-Network/blockchain-node/pull/402

**Pull request closed without merging**